src/Rematch.jl

```diff
@@ -19,6 +19,14 @@ function assert_num_fields(::Type{T}, matched::Integer) where {T}
     @assert actual == matched "Tried to match $matched fields of $actual field struct $T"
 end
 
+"""
+Statically get the fieldcount of a type. Useful to avoid runtime calls to
+fieldcount.
+"""
+@generated function evaluated_fieldcount(t::Type{T}) where T
+    fieldcount(T)
+end
```

Comment:
Now that we have more context on this whole `@generated` discussion, I wonder if this needs to be `@generated`, or if the whole function needs to exist at all? I would think that since it only depends on static information, `fieldcount(T)` should compile away into a single Int value, but I don't see that in my simple tests at the REPL... I'll ask about this on julia slack... it's weird though; the fieldcount shouldn't be able to change, so i don't get it.

(No comment here other than a TODO(nathan.daly).)